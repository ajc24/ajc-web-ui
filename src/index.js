/* Export all of the header components */
export { default as Header } from './ui-elements/header/Header';

/* Export all of the icon components */
export { default as SmallDownArrowIcon } from './ui-elements/icons/SmallDownArrowIcon';
export { default as SmallRightArrowIcon } from './ui-elements/icons/SmallRightArrowIcon';
