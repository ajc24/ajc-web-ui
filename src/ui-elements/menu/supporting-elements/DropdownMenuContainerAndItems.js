import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import SmallRightArrowIcon from '../../icons/SmallRightArrowIcon';
import '../../../css/common.css';
import '../../../css/menu.css';

class DropdownMenuContainerAndItems extends React.Component {
  constructor(props) {
    super(props);
    this.setDropdownItemsListXPosition = this.setDropdownItemsListXPosition.bind(this);
    this.setDropdownItemsListYPosition = this.setDropdownItemsListYPosition.bind(this);
  }

  componentDidMount() {
    setTimeout(() => {
      this.setDropdownItemsListYPosition();
      this.setDropdownItemsListXPosition();
      window.addEventListener('resize', this.setDropdownItemsListXPosition, { capture: true });
    }, 10);
  }

  componentWillUnmount() {
    /* Remove the dynamically added resize event when the component is unmounted */
    window.removeEventListener('resize', this.setDropdownItemsListXPosition, { capture: true });
  }

  /**
   * Sets the X position of the dropdown menu items list container
   * The X position is always re-adjustable since screen resizes will affect the X position of the container
   */
  setDropdownItemsListXPosition() {
    /* Find the boundaries of the main menu item for this component */
    const mainMenuItemElement = document.querySelector(`div[id="${this.props.parentId}"]`);
    const mainMenuItemBoundary = mainMenuItemElement.getBoundingClientRect();
    /* Find the dropdown items list root element and set the width */
    const itemsListRootElement = document.querySelector(`div[id="${this.props.id}"]`);
    const itemsListRootBoundary = itemsListRootElement.getBoundingClientRect();
    if (itemsListRootBoundary.width < mainMenuItemBoundary.width) {
      itemsListRootElement.style.width = `${mainMenuItemBoundary.width}px`;
    } else {
      itemsListRootElement.style.width = `${itemsListRootBoundary.width}px`;
    }
    itemsListRootElement.style.minWidth = itemsListRootElement.style.width;
    /* Determine if the leftmost position of the dropdown menu items container needs to be adjusted to fit the screen size */
    let leftmostPosition = mainMenuItemBoundary.left;
    if (mainMenuItemBoundary.left + itemsListRootBoundary.width > window.innerWidth) {
      leftmostPosition = window.innerWidth - itemsListRootBoundary.width;
    }
    itemsListRootElement.style.left = `${leftmostPosition}px`;
  }

  /**
   * Sets the Y position of the dropdown menu items list container
   * This only needs to be set once since screen resizes will only mainly affect the X axis positioning of the container
   */
  setDropdownItemsListYPosition() {
    /* Ensure the window is scrolled to the top before marking the position of the dropdown items list */
    window.scrollTo({
      top: 0,
      behavior: 'smooth'
    });
    /* Find the boundaries of the main menu item for this component */
    const mainMenuItemElement = document.querySelector(`div[id="${this.props.parentId}"]`);
    const mainMenuItemBottomPos = mainMenuItemElement.getBoundingClientRect().bottom;
    /* Find the dropdown items list root element and set the top position */
    const itemsListRootElement = document.querySelector(`div[id="${this.props.id}"]`);
    itemsListRootElement.style.top = `${mainMenuItemBottomPos}px`;
  }

  render() {
    let rootStyling = 'ajc-flex-column ajc-dropdown-menu';
    let dropdownMenuItemStyling = 'ajc-flex-row ajc-dropdown-menu-item ajc-font-menu';
    if (this.props.colour === 'red') {
      rootStyling += ' ajc-background-red';
      dropdownMenuItemStyling += ' ajc-menu-item-red';
    } else {
      rootStyling += ' ajc-background-grey-2';
      dropdownMenuItemStyling += ' ajc-menu-item-grey';
    }
    if (this.props.isDisplayed === true) {
      rootStyling += ' ajc-dropdown-menu-visible ajc-visibility-visible';
    } else {
      rootStyling += ' ajc-dropdown-menu-hidden ajc-visibility-hidden';
    }
    return (
      <div id={this.props.id} className={rootStyling}>
        {
          this.props.dropdownMenuItemsList.map((menuItem, index) => {
            return <Link to={menuItem.route} key={`${this.props.id}--key-menu-item-${index}`} className={dropdownMenuItemStyling} id={`${this.props.id}--dropdown-menu-item-${index}`} tabIndex={0}>
              &nbsp;&nbsp;{menuItem.title}&nbsp;&nbsp;
              <SmallRightArrowIcon colour="white" />
              &nbsp;&nbsp;
            </Link>
          })
        }
      </div>
    );
  }
}
DropdownMenuContainerAndItems.propTypes = {
  /** The components unique identifier. */
  id: PropTypes.string.isRequired,
  /** The background colour for the container. By default this is set to grey. */
  colour: PropTypes.oneOf([ 'grey', 'red' ]),
  /** The unique identifier of this components parent DropdownMenu component. */
  parentId: PropTypes.string.isRequired,
  /** Specifies whether the dropdown menu container is visible or hidden in the UI. */
  isDisplayed: PropTypes.bool.isRequired,
  /** The dropdown menu data which defines the list of dropdown menu items and defines the route and title for each individual item. */
  dropdownMenuItemsList: PropTypes.arrayOf(PropTypes.shape({
    route: PropTypes.string,
    title: PropTypes.string
  })).isRequired,
};
DropdownMenuContainerAndItems.defaultProps = {
  colour: 'grey'
};
export default DropdownMenuContainerAndItems;
