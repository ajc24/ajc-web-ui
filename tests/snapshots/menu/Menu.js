import 'jsdom-global/register';
import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { TestDev } from 'ajc-jest-enzyme';
import { Menu } from '../../../src';
import DropdownMenuContainerAndItems from '../../../src/ui-elements/menu/supporting-elements/DropdownMenuContainerAndItems';

describe('Menu', () => {
  /* Create the test data for the menu */
  let componentDidMountSpy;
  const testMenuItemsList = [
    {
      route: '/test-item-1',
      title: 'Menu Item 1',
    },
    {
      title: 'Dropdown Menu Item 1',
      dropdownMenuItemsList: [
        {
          route: '/test-dropdown-item-1-1',
          title: 'Dropdown Menu Item 1.1',
        },
        {
          route: '/test-dropdown-item-1-2',
          title: 'Dropdown Menu Item 1.2',
        },
      ],
    },
    {
      route: '/test-item-2',
      title: 'Menu Item 2',
    },
    {
      title: 'Dropdown Menu Item 2',
      dropdownMenuItemsList: [
        {
          route: '/test-dropdown-item-2-1',
          title: 'Dropdown Menu Item 2.1',
        },
        {
          route: '/test-dropdown-item-2-2',
          title: 'Dropdown Menu Item 2.2',
        },
      ],
    },
  ];

  beforeAll(() => {
    componentDidMountSpy = jest
      .spyOn(DropdownMenuContainerAndItems.prototype, 'componentDidMount')
      .mockImplementation(() => {});
  });

  afterAll(() => {
    componentDidMountSpy.mockRestore();
  });

  describe('Default props and rendering', () => {
    let jsonSnapshot;

    beforeAll(() => {
      jsonSnapshot = TestDev.createSnapshot(
        <React.Fragment>
          <BrowserRouter>
            <Menu id="test-menu" menuItemsList={testMenuItemsList} />
          </BrowserRouter>
        </React.Fragment>
      );
    });
    
    it('verifies the snapshot for the component', () => {
      expect(jsonSnapshot).toMatchSnapshot();
    });
  });

  describe('Transferred props and rendering: Component with red background', () => {
    let jsonSnapshot;

    beforeAll(() => {
      jsonSnapshot = TestDev.createSnapshot(
        <React.Fragment>
          <BrowserRouter>
            <Menu id="test-menu" menuItemsList={testMenuItemsList} colour="red" />
          </BrowserRouter>
        </React.Fragment>
      );
    });

    it('verifies the snapshot for the component', () => {
      expect(jsonSnapshot).toMatchSnapshot();
    });
  });
});
