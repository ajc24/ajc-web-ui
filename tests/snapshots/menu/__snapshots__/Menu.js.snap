// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Default props and rendering verifies the snapshot for the component 1`] = `
<div
  className="ajc-flex-row ajc-flex-justify-center ajc-menu ajc-width-4k ajc-background-grey-2"
  id="test-menu-root"
  role="navigation"
>
  <div
    className="ajc-flex-row ajc-flex-justify-center ajc-width-full-hd"
  >
    <div
      className="ajc-flex-row ajc-flex-justify-center ajc-width-hd"
    >
      <div
        className="ajc-flex-row ajc-flex-justify-center ajc-width-hd-with-padding ajc-menu-content"
        id="test-menu"
      >
        <div
          className="ajc-flex-row ajc-justify-left"
          id="menu-item-0-root"
        >
          <div
            className="ajc-flex-row ajc-menu-item ajc-menu-item-grey"
            id="menu-item-0"
          >
            <a
              className="ajc-flex-row ajc-menu-item-content ajc-font-menu ajc-text-white"
              href="/test-item-1"
              onClick={[Function]}
              tabIndex={0}
            >
                
              Menu Item 1
                
            </a>
          </div>
        </div>
        <div
          className="ajc-flex-row ajc-justify-left ajc-menu-item"
          id="dropdown-menu-1"
        >
          <div
            className="ajc-flex-row ajc-menu-item-content ajc-font-menu ajc-text-white ajc-menu-item-grey"
            data-expanded="false"
            onClick={[Function]}
            onKeyPress={[Function]}
            tabIndex={0}
          >
              
            Dropdown Menu Item 1
              
            <div
              className="ajc-flex-row ajc-justify-left"
              role="presentation"
            >
              <div
                className="ajc-flex-row ajc-small-down-arrow-icon ajc-border-colour-white"
              />
            </div>
              
          </div>
          <div
            className="ajc-flex-column ajc-dropdown-menu ajc-background-grey-2 ajc-dropdown-menu-hidden ajc-visibility-hidden"
            id="dropdown-menu-1--content-parent"
          >
            <a
              className="ajc-flex-row ajc-dropdown-menu-item ajc-font-menu ajc-menu-item-grey"
              href="/test-dropdown-item-1-1"
              id="dropdown-menu-1--content-parent--dropdown-menu-item-0"
              onClick={[Function]}
              tabIndex={0}
            >
                
              Dropdown Menu Item 1.1
                
              <div
                className="ajc-flex-row ajc-justify-left"
                role="presentation"
              >
                <div
                  className="ajc-flex-row ajc-small-right-arrow-icon ajc-border-colour-white"
                />
              </div>
                
            </a>
            <a
              className="ajc-flex-row ajc-dropdown-menu-item ajc-font-menu ajc-menu-item-grey"
              href="/test-dropdown-item-1-2"
              id="dropdown-menu-1--content-parent--dropdown-menu-item-1"
              onClick={[Function]}
              tabIndex={0}
            >
                
              Dropdown Menu Item 1.2
                
              <div
                className="ajc-flex-row ajc-justify-left"
                role="presentation"
              >
                <div
                  className="ajc-flex-row ajc-small-right-arrow-icon ajc-border-colour-white"
                />
              </div>
                
            </a>
          </div>
        </div>
        <div
          className="ajc-flex-row ajc-justify-left"
          id="menu-item-2-root"
        >
          <div
            className="ajc-flex-row ajc-menu-item ajc-menu-item-grey"
            id="menu-item-2"
          >
            <a
              className="ajc-flex-row ajc-menu-item-content ajc-font-menu ajc-text-white"
              href="/test-item-2"
              onClick={[Function]}
              tabIndex={0}
            >
                
              Menu Item 2
                
            </a>
          </div>
        </div>
        <div
          className="ajc-flex-row ajc-justify-left ajc-menu-item"
          id="dropdown-menu-3"
        >
          <div
            className="ajc-flex-row ajc-menu-item-content ajc-font-menu ajc-text-white ajc-menu-item-grey"
            data-expanded="false"
            onClick={[Function]}
            onKeyPress={[Function]}
            tabIndex={0}
          >
              
            Dropdown Menu Item 2
              
            <div
              className="ajc-flex-row ajc-justify-left"
              role="presentation"
            >
              <div
                className="ajc-flex-row ajc-small-down-arrow-icon ajc-border-colour-white"
              />
            </div>
              
          </div>
          <div
            className="ajc-flex-column ajc-dropdown-menu ajc-background-grey-2 ajc-dropdown-menu-hidden ajc-visibility-hidden"
            id="dropdown-menu-3--content-parent"
          >
            <a
              className="ajc-flex-row ajc-dropdown-menu-item ajc-font-menu ajc-menu-item-grey"
              href="/test-dropdown-item-2-1"
              id="dropdown-menu-3--content-parent--dropdown-menu-item-0"
              onClick={[Function]}
              tabIndex={0}
            >
                
              Dropdown Menu Item 2.1
                
              <div
                className="ajc-flex-row ajc-justify-left"
                role="presentation"
              >
                <div
                  className="ajc-flex-row ajc-small-right-arrow-icon ajc-border-colour-white"
                />
              </div>
                
            </a>
            <a
              className="ajc-flex-row ajc-dropdown-menu-item ajc-font-menu ajc-menu-item-grey"
              href="/test-dropdown-item-2-2"
              id="dropdown-menu-3--content-parent--dropdown-menu-item-1"
              onClick={[Function]}
              tabIndex={0}
            >
                
              Dropdown Menu Item 2.2
                
              <div
                className="ajc-flex-row ajc-justify-left"
                role="presentation"
              >
                <div
                  className="ajc-flex-row ajc-small-right-arrow-icon ajc-border-colour-white"
                />
              </div>
                
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Menu Transferred props and rendering: Component with red background verifies the snapshot for the component 1`] = `
<div
  className="ajc-flex-row ajc-flex-justify-center ajc-menu ajc-width-4k ajc-background-red"
  id="test-menu-root"
  role="navigation"
>
  <div
    className="ajc-flex-row ajc-flex-justify-center ajc-width-full-hd"
  >
    <div
      className="ajc-flex-row ajc-flex-justify-center ajc-width-hd"
    >
      <div
        className="ajc-flex-row ajc-flex-justify-center ajc-width-hd-with-padding ajc-menu-content"
        id="test-menu"
      >
        <div
          className="ajc-flex-row ajc-justify-left"
          id="menu-item-0-root"
        >
          <div
            className="ajc-flex-row ajc-menu-item ajc-menu-item-red"
            id="menu-item-0"
          >
            <a
              className="ajc-flex-row ajc-menu-item-content ajc-font-menu ajc-text-white"
              href="/test-item-1"
              onClick={[Function]}
              tabIndex={0}
            >
                
              Menu Item 1
                
            </a>
          </div>
        </div>
        <div
          className="ajc-flex-row ajc-justify-left ajc-menu-item"
          id="dropdown-menu-1"
        >
          <div
            className="ajc-flex-row ajc-menu-item-content ajc-font-menu ajc-text-white ajc-menu-item-red"
            data-expanded="false"
            onClick={[Function]}
            onKeyPress={[Function]}
            tabIndex={0}
          >
              
            Dropdown Menu Item 1
              
            <div
              className="ajc-flex-row ajc-justify-left"
              role="presentation"
            >
              <div
                className="ajc-flex-row ajc-small-down-arrow-icon ajc-border-colour-white"
              />
            </div>
              
          </div>
          <div
            className="ajc-flex-column ajc-dropdown-menu ajc-background-red ajc-dropdown-menu-hidden ajc-visibility-hidden"
            id="dropdown-menu-1--content-parent"
          >
            <a
              className="ajc-flex-row ajc-dropdown-menu-item ajc-font-menu ajc-menu-item-red"
              href="/test-dropdown-item-1-1"
              id="dropdown-menu-1--content-parent--dropdown-menu-item-0"
              onClick={[Function]}
              tabIndex={0}
            >
                
              Dropdown Menu Item 1.1
                
              <div
                className="ajc-flex-row ajc-justify-left"
                role="presentation"
              >
                <div
                  className="ajc-flex-row ajc-small-right-arrow-icon ajc-border-colour-white"
                />
              </div>
                
            </a>
            <a
              className="ajc-flex-row ajc-dropdown-menu-item ajc-font-menu ajc-menu-item-red"
              href="/test-dropdown-item-1-2"
              id="dropdown-menu-1--content-parent--dropdown-menu-item-1"
              onClick={[Function]}
              tabIndex={0}
            >
                
              Dropdown Menu Item 1.2
                
              <div
                className="ajc-flex-row ajc-justify-left"
                role="presentation"
              >
                <div
                  className="ajc-flex-row ajc-small-right-arrow-icon ajc-border-colour-white"
                />
              </div>
                
            </a>
          </div>
        </div>
        <div
          className="ajc-flex-row ajc-justify-left"
          id="menu-item-2-root"
        >
          <div
            className="ajc-flex-row ajc-menu-item ajc-menu-item-red"
            id="menu-item-2"
          >
            <a
              className="ajc-flex-row ajc-menu-item-content ajc-font-menu ajc-text-white"
              href="/test-item-2"
              onClick={[Function]}
              tabIndex={0}
            >
                
              Menu Item 2
                
            </a>
          </div>
        </div>
        <div
          className="ajc-flex-row ajc-justify-left ajc-menu-item"
          id="dropdown-menu-3"
        >
          <div
            className="ajc-flex-row ajc-menu-item-content ajc-font-menu ajc-text-white ajc-menu-item-red"
            data-expanded="false"
            onClick={[Function]}
            onKeyPress={[Function]}
            tabIndex={0}
          >
              
            Dropdown Menu Item 2
              
            <div
              className="ajc-flex-row ajc-justify-left"
              role="presentation"
            >
              <div
                className="ajc-flex-row ajc-small-down-arrow-icon ajc-border-colour-white"
              />
            </div>
              
          </div>
          <div
            className="ajc-flex-column ajc-dropdown-menu ajc-background-red ajc-dropdown-menu-hidden ajc-visibility-hidden"
            id="dropdown-menu-3--content-parent"
          >
            <a
              className="ajc-flex-row ajc-dropdown-menu-item ajc-font-menu ajc-menu-item-red"
              href="/test-dropdown-item-2-1"
              id="dropdown-menu-3--content-parent--dropdown-menu-item-0"
              onClick={[Function]}
              tabIndex={0}
            >
                
              Dropdown Menu Item 2.1
                
              <div
                className="ajc-flex-row ajc-justify-left"
                role="presentation"
              >
                <div
                  className="ajc-flex-row ajc-small-right-arrow-icon ajc-border-colour-white"
                />
              </div>
                
            </a>
            <a
              className="ajc-flex-row ajc-dropdown-menu-item ajc-font-menu ajc-menu-item-red"
              href="/test-dropdown-item-2-2"
              id="dropdown-menu-3--content-parent--dropdown-menu-item-1"
              onClick={[Function]}
              tabIndex={0}
            >
                
              Dropdown Menu Item 2.2
                
              <div
                className="ajc-flex-row ajc-justify-left"
                role="presentation"
              >
                <div
                  className="ajc-flex-row ajc-small-right-arrow-icon ajc-border-colour-white"
                />
              </div>
                
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
